enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+VIP functionality that will have to be a separate operation from upgrade.
+
+### Version Skew Strategy
+
+This will also be handled as it is today. The keepalived configuration used
+for the new version must work with the old version until all nodes have been
+upgraded. If breaking changes are needed, there will be a migration mechanism
+that runs after upgrade is complete (similar to what we did in the multicast
+to unicast migration).
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
+
+#### Support Procedures
+
+NA
+
+## Implementation History
+
+4.11: Initial implementation
+
+## Drawbacks
+
+Nothing significant. A very small amount of compute resources will be used
+to manage the new VIPs. Even this can be avoided by simply not specifying
+the second VIP if it is not needed.
+
+## Alternatives
+
+Implement this somewhat like the
+[existing workaround](https://github.com/yboaron/kepalived-v6) - create a
+separate instance of keepalived that is responsible for ipv6 only. This
+results in some duplication of containers and configs, but it would probably
+simplify the code changes.
+
+### Alternative install-config layouts
+
+#### Just add v6 versions of the VIP fields
+```yaml
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIP: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is a minimal change, but it leaves us with somewhat of a problem in that
+we can't assume the old fields always contain a v4 address since previous
+releases allowed either v4 or v6. We may be able to deal with that by
+detecting the appropriate ip version for the old fields and populating new
+fields on the api side. E.g. if apiVIP is v4, we populate
+`platformStatus.apiServerInternalIPv4` and if it is v6 we populate
+`apiServerInternalIPv6` instead. That way the only place we need to worry about
+the difference is in the installer. We would need a validation to ensure a v6
+VIP isn't specified in both fields.
+
+Note that this assumes we replace the existing api fields with two new ones
+per VIP and do not reuse the existing ones at all. I think we're going to want
+to do that regardless of what install-config layout we use so we don't need
+logic on the backend in machine-config-operator and baremetal-runtimecfg to
+handle older cluster configs.
+
+#### Deprecate the old fields and add new v4 and v6 fields
+```yaml
+platform:
+  baremetal:
+    # Deprecated
+    apiVIP: 192.168.111.5
+    # Deprecated
+    ingressVIP: 192.168.111.4
+
+    apiVIPv4: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIPv4: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is more of a 1-1 mapping to what I anticipate the platformStatus api
+will look like. We would still need compatibility logic to handle old-style
+install-configs that set the deprecated fields, but having no overlap between
+the old and new-style configs might simplify the logic somewhat.
+
+#### Add plural fields that take a list of VIPs
```

Comment:
I'm wondering if this data structure wouldn't be the most in-line with other network parameters configurable nowadays. We have already `clusterNetworks`, `serviceNetworks` and so on that are plural, what does not necessarily mean that they accept every single combination.

When in Assisted Installer we were implementing dual-stack support, we had similar concerns - whether fields that in single-stack accept 1 entry and in dual-stack exactly 2 entries deserve to be lists or should we take the simplest path by introducing `-secondary` suffix. We picked the former as a way of keeping the configuration elegant for the price of internal validations to be a bit more complex.

Reply:
Dan mentioned the same thing in his comments on a previous revision and if we were doing a clean sheet design I'd absolutely prefer this. With the existing config though, I think there's also an elegance to not having multiple fields that configure the same thing. Yes, we can handle that, but is this enough better to justify the complication of deprecating the old field and complicating the backend implementation? I don't think so, but I don't have strong feelings about this. Most of the options presented here are perfectly valid (except the one Kiran noted above) so I'm open to the alternatives if others feel strongly about it.

Reply:
This alternative is my preference. It keeps it consistent with other fields. Deprecation is not terribly complex to achieve. As discussed above, the backend implementation should not be effected by deprecated fields.